### ISSUE TITLE
WinUI 3 - The attachable property 'InputBindings' was not found in type 'Window'. 

### ISSUE THREAD

**Issue by tuffstuff9** [giosali/HotkeyUtility#2]
Hi, I am trying to figure out how to use this utility in a WinUI 3 application. I followed the quickstart guide and added this in my XAML view. 
![image](https://user-images.githubusercontent.com/57072903/179764374-ec4c275e-3603-43c5-b17b-40812b6316b7.png)
However, I am getting the error in the title. I found this thread (https://github.com/microsoft/WindowsAppSDK/discussions/1974) which contains dead links. I used the wayback machine to view those links and I'm still not sure how to get this working. Any assistance would be much appreciated. 

Here are the two archived links in that thread for your reference:
https://web.archive.org/web/20211223095955/https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.window?view=winui-3.0
https://web.archive.org/web/20211223084314/https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.page?view=winui-3.0

**Comment by giosali** [comment]
Hmm, I'm not all that familiar with WinUI 3 to be honest. I briefly looked into it when I was working on modernizing a WPF application but I don't remember too much about it.

Is WinUI 3 its own stand-alone framework or is it something that you integrated into an existing UI framework (e.g. WinForms, WPF, UWP, etc.)?

I think the issue is that, at the moment, `HotkeyUtility` is only intended to be used in a WPF application. If you look at the [Window class](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window?view=windowsdesktop-6.0) in a WPF application, it contains a property called [InputBindings](https://docs.microsoft.com/en-us/dotnet/api/system.windows.uielement.inputbindings?view=windowsdesktop-6.0#system-windows-uielement-inputbindings=). In your case, it looks like you might be using the [Window class](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window?view=windows-app-sdk-1.1) that belongs to the `Microsoft.UI.Xaml` namespace which doesn't contain an `InputBindings` property.

**Comment by tuffstuff9** [comment]
Thanks for your reply. WinUI 3 is indeed a standalone UI framework. It is XAML based, much like WPF, and intended to be the modern choice for windows development (instead of UWP, WPF). I was looking for a hotkey manager and yours looks like the most capable, which is also intended for WPF (the others were built around winforms and aren’t viable for winui 3). There is a lot of carryover between WPF and WinUI3, so I assumed a wpf hotkey manager would work with winui3. 

You are correct about your assumption. Does your library require itself to partly be implemented in the view? If so, is there an alternative to using the InputBindings property? Perhaps this property has been updated into something else in winui3?

**Comment by giosali** [comment]
> Does your library require itself to partly be implemented in the view?

Actually, you don't have to implement it in the view. You can also implement it in the code-behind/view model if you like:

```csharp
using System;
using System.Windows.Input;
using HotkeyUtility;

protected override void OnInitialized(EventArgs e)
{
    Hotkey hotkey = new(Key.Space, ModifierKeys.Alt, HotkeyBinding_Pressed);
    HotkeyManager hotkeyManager = HotkeyManager.GetHotkeyManager();
    _ = hotkeyManager.TryAddHotkey(hotkey);
}

public void HotkeyBinding_Pressed(object sender, HotkeyEventArgs e)
{
    // Handle hotkey press here
}
```

Since you're using WinUI 3, I'm not sure if you'll have an [OnInitialized](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.oninitialized?view=windowsdesktop-6.0) method to override but I'm sure something similar exists.

**Comment by tuffstuff9** [comment]
Thank you for that. I have managed to get it working in my viewmodel. The singleton pattern really makes everything a lot easier. I was wondering if there was any way to record keypresses in a textbox to assign to a hotkey during runtime?

**Comment by giosali** [comment]
> I was wondering if there was any way to record keypresses in a textbox to assign to a hotkey during runtime?

I actually implemented something exactly like that in [my application](https://github.com/giosali/reginald) I mentioned earlier.

![reginald](https://user-images.githubusercontent.com/70548443/180044990-f8f5f5ba-5789-46b9-aa50-495d22fb305d.png)

The way you would handle it is by monitoring the `PreviewKeyDown` event on a `TextBox` control and using the keys the user types as the new hotkey.

Here's a slightly modified version of how I implemented it in my application:

```csharp
// You're going to need to keep track of the current hotkey combination to replace it later.
private Hotkey MyHotkey { get; set; }

public void HotkeyTextBox_PreviewKeyDown(object sender, KeyEventArgs e)
{
    ModifierKeys modifiers = Keyboard.Modifiers;
    switch (modifiers)
    {
        // Selected when a modifier key was pressed (other than the Windows logo key).
        case not ModifierKeys.None when modifiers != ModifierKeys.Windows:
            Key key;
            switch (key = e.Key == Key.System ? e.SystemKey : e.Key)
            {
                // Selected when a key other than Shift, Control, or Alt is pressed.
                case < Key.LeftShift:
                case > Key.RightAlt:
                    // The F12 key is reserved for use by the debugger at all times, so it should not be registered as a hotkey. Even when you are not debugging an application, F12 is reserved in case a kernel-mode debugger or a just-in-time debugger is resident.
                    if (key == Key.F12)
                    {
                        break;
                    }

                    // Ensures that if the user tries registering a key binding that's
                    // clearly already registered, the request will simply be ignored.
                    if (key == MyHotkey.Key && modifiers == MyHotkey.Modifiers)
                    {
                        break;
                    }

                    HotkeyManager hotkeyManager = HotkeyManager.GetHotkeyManager();
                    
                    // This checks to see if the hotkey binding you're replacing actually exists.
                    Hotkey hotkey = hotkeyManager.GetHotkeys().Find(MyHotkey.Key, MyHotkey.Modifiers);
                    if (hotkey is not null)
                    {
                        // If it does exist, we replace it here:
                        _ = TryReplaceHotkey(MyHotkey.Key, MyHotkey.Modifiers, key, modifiers);
                        
                        // Keep track of the new hotkey binding
                        MyHotkey.Key = key;
                        MyHotkey.Modifiers = modifiers;
                    }

                    break;
            }

            break;
    }
}
```

That's it pretty much. You're probably going to have to manually update the text in the `TextBox`. You could probably have a property like:

```csharp
public string HotkeyTextBoxInput { get; set; }
```

...and bind that to the `Text` property of the `TextBox` control. When you successfully replace the old hotkey, convert `MyHotkey.Key` and `MyHotkey.Modifiers` into a string representation and assign it to `HotkeyTextBoxInput`. This is a method I created to do that that you can use if you like:

```csharp
private static string ConvertKeysToString(Key key, ModifierKeys modifiers)
{
    return modifiers.ToString().Replace(",", " +") + " + " + key.ToString();
}
```

**Comment by tuffstuff9** [comment]
Thank you for this. I ran into some trouble when using this for WinUI 3, but I have managed to resolve it after a lot of trial and error. 

WinUI 3's PreviewKeyDown has an event of type `KeyRoutedEventArgs` rather than `KeyEventArgs`. Therefore, the type of `e.Key` is a Windows.System.VirtualKey enum rather than a System.Windows.Input.Key enum which is what your library requires. In order to fix this, the VirtualKey must be converted to a Key. This can be done by:
```cs
Key key = KeyInterop.KeyFromVirtualKey((int)e.Key);
```

Have you considered the case where the user wants to create a new hotkey during runtime? For example, if they want to create a new hotkey to do a math sum. I have tried to create a blank hotkey, but when I access MyHotkey.Key, it is null and I can not assign anything to it. 

My thought process is that I can create a new blank hotkey in a click handler (like a button to add a new hotkey in a list) like this: 
```cs
private void Button_Click(object sender, RoutedEventArgs e)
    {
        Hotkey hotkey = new(Key.None, ModifierKeys.None, ActivateHotkey(int.Parse(num1TextBox.Text), int.Parse(num2TextBox.Text)));
    }

private void ActivateHotkey(int num1, int num2)
    {
        Debug.WriteLine(num1 + num2);
    }
```

~~However, the hotkey does not accept the `ActivateHotkey` function. Is another way of doing this to perhaps add num1 and num2 as a property to the Hotkey class, and store those during creation, so that they can be modified later too, and manipulated in the hotkeypressed function?~~

EDIT: I have come up with a way to do this. If I edit the Hotkey.cs class and add my desired properties, like num1 and num2, I can cast the sender object into a hotkey and then access hotkey.num1 in the hotkeypressed method. I am now working on a way to distinguish between left and right modifier keys, because it seems that ModifierKeys does not distinguish between the locations. Then, I will have to see how hotkeys and their num1 and num2 can be edited during runtime. 

EDIT2: Also, it seems that HotkeyManager has no method called Find(), unless I'm missing something. 
![image](https://user-images.githubusercontent.com/57072903/180264388-3c356c54-2916-4ba1-afdf-5d183ebf5abc.png)
